docs/proposals/policy.md

```diff
@@ -0,0 +1,389 @@
+In order to determine whether a request is authorized, the AuthorizationAttributes are tested in the following order:
+  1. all deny RoleBinding PolicyRules in the global namespace - short circuit on match
+  1. all allow RoleBinding PolicyRules in the global namespace - short circuit on match
+  1. all deny RoleBinding PolicyRules in the namespace - short circuit on match
+  1. all allow RoleBinding PolicyRules in the namespace - short circuit on match
+  1. deny by default
+
+### Configtime Authorization Types
+These are the types used in the policy example above.  They allow us: to quickly find the policy rules that apply to namespace in etcd, to easily delegate project level control to a project admin, and to easily segregate all role control (a project admin can create/update/delete RoleBindings, but not Roles).
+```
+// PolicyRule holds information that describes a policy rule, but does not contain information 
+// about who the rule applies to or which namespace the rule applies to.
+type PolicyRule struct {
+	// Deny is true if any request matching this rule should be denied.  If false, any request matching this rule is allowed.
+	Deny bool
+	// Verbs is a list of Verbs that apply to ALL the ResourceKinds and AttributeRestrictions contained in this rule.  "*" represents all kinds.
+	Verbs []Verb
+	// AttributeRestrictions will vary depending on what the Authorizer/AuthorizationAttributeBuilder pair supports.
+	// If the Authorizer does not recognize how to handle the AttributeRestrictions, the Authorizer should report an error.
+	AttributeRestrictions runtime.EmbeddedObject
+	// ResourceKinds is a list of kinds this rule applies to.  "*" represents all kinds.
+	ResourceKinds []string
+}
+
+// Role is a logical grouping of PolicyRules that can be referenced as a unit by RoleBindings.
+type Role struct {
+	kapi.TypeMeta
+	kapi.ObjectMeta
+
+	Rules []PolicyRule
+}
+
+// RoleBinding references a Role, but not contain it.  It adds who and namespace information.  
+// It can reference any Role in the same namespace or in the global namespace.
+type RoleBinding struct {
+	kapi.TypeMeta
+	kapi.ObjectMeta
+
+	UserNames    []string
+	GroupNames    []string
+
+	// Since Policy is a singleton, this is sufficient knowledge to locate a role
+	// RoleRefs can only reference the current namespace and the global namespace
+	// If the RoleRef cannot be resolved, the Authorizer must return an error.
+	RoleRef kapi.ObjectReference
+}
+
+// Policy is a object that holds all the Roles and RoleBindings for a particular namespace.  There is at most
+// one Policy document per namespace.
+// It is an attempt to mitigate referential integrity problems, but it is worth noting that it does not solve them.
+// Referential integrity problems still exist for references to the global namespace.  If that problem is resolved
+// then it seems more reasonable to destroy this resource and use the referential integrity solution to solve
+// all cases.
+type Policy struct{
```

Comment:
I don't think Policy should contain Roles and Bindings.  I think bindings are separate, and bindings apply to a single policy via some convention.  The reason being is that you don't need atomic integrity across role definitions, you just need atomic integrity across bindings.  If a binding points to a role that does not exist, the permissions granted by that role are gone and you default to no policy.

Also, we want to reuse Policies across multiple projects, so coupling them with bindings means that they are not reusable.